### PR TITLE
SNAP-2467

### PR DESCRIPTION
**Pull request by ymahajan**
## Changes proposed in this pull request

Added support for default column value in alter table add column

## Patch testing

precheckin, manual testing 

## ReleaseNotes.txt changes

None

## Other PRs 

None

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/StoreCallbacksImpl.scala [hunk 1]
```diff
@@ -522,7 +522,7 @@ object StoreCallbacksImpl extends StoreCallbacks with Logging with Serializable
         val columnNullable = context.getColumnNullable
         logDebug(s"StoreCallbacksImpl.performConnectorOp alter table ")
         session.alterTable(tableName, addOrDropCol, StructField(columnName,
-          CatalystSqlParser.parseDataType(columnDataType), columnNullable))
+          CatalystSqlParser.parseDataType(columnDataType), columnNullable), defaultValue = "")
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/StoreCallbacksImpl.scala]: smart connector route also needs to add defaultValue argument to the procedure

File: core/src/main/scala/org/apache/spark/sql/SnappySession.scala [hunk 2]
```diff
@@ -1470,7 +1471,7 @@ class SnappySession(_sc: SparkContext) extends SparkSession(_sc) {
   }
 
   private[sql] def alterTable(tableIdent: QualifiedTableName, isAddColumn: Boolean,
-      column: StructField): Unit = {
+      column: StructField, defaultValue: String): Unit = {
     val plan = try {
       sessionCatalog.lookupRelation(tableIdent)
     } catch {
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/SnappySession.scala]: No implementation?

**Comment by ymahajan** [comment]
@sumwale Thanks,  I have a few changes in my local checkout which I am planning to commit today.

File: core/src/main/scala/org/apache/spark/sql/SnappyContext.scala [hunk 3]
```diff
@@ -144,10 +144,11 @@ class SnappyContext protected[spark](val snappySession: SnappySession)
     * @param tableName
     * @param isAddColumn
     * @param column
+    * @param defaultValue
     */
   def alterTable(tableName: String, isAddColumn: Boolean,
-                 column: StructField): Unit = {
-    snappySession.alterTable(tableName, isAddColumn, column)
+                 column: StructField, defaultValue: String = ""): Unit = {
+    snappySession.alterTable(tableName, isAddColumn, column, defaultValue)
```
[the remark this answers is not in the document]
Comment (dshirish) [reply core/src/main/scala/org/apache/spark/sql/SnappyContext.scala]: Made the suggested changes

File: core/src/main/scala/org/apache/spark/sql/SnappyDDLParser.scala [hunk 4]
```diff
@@ -307,6 +309,14 @@ abstract class SnappyDDLParser(session: SparkSession)
     }
   }
 
+  protected final def defaultVal: Rule1[String] = rule {
+    (DEFAULT).? ~> ((defVal: Any) =>
+      defVal match {
+        case Some(v) => v.asInstanceOf[String].trim
+        case None => SnappyParserConsts.SELECT.upper
+      })
+  }
```
[the remark this answers is not in the document]
Comment (dshirish) [reply core/src/main/scala/org/apache/spark/sql/SnappyDDLParser.scala]: Made the suggested changes

File: core/src/main/scala/org/apache/spark/sql/SnappyDDLParser.scala [hunk 5]
```diff
@@ -493,7 +503,7 @@ abstract class SnappyDDLParser(session: SparkSession)
 
   protected def alterTable: Rule1[LogicalPlan] = rule {
     ALTER ~ TABLE ~ tableIdentifier ~ (
-        ADD ~ COLUMN.? ~ column ~ EOI ~> AlterTableAddColumn |
+        ADD ~ COLUMN.? ~ column ~ DEFAULT.? ~ defaultVal ~ EOI ~> AlterTableAddColumn |
```
[the remark this answers is not in the document]
Comment (dshirish) [reply core/src/main/scala/org/apache/spark/sql/SnappyDDLParser.scala]: Corrected the code

File: core/src/main/scala/org/apache/spark/sql/SnappySession.scala [hunk 6]
```diff
@@ -1428,14 +1428,15 @@ class SnappySession(_sc: SparkContext) extends SparkSession(_sc) {
   }
 
   private[sql] def alterTable(tableName: String, isAddColumn: Boolean,
-      column: StructField): Unit = {
+      column: StructField, defaultValue: String): Unit = {
     val qualifiedTable = sessionCatalog.newQualifiedTableName(tableName)
     if (sessionCatalog.caseSensitiveAnalysis) {
-      alterTable(qualifiedTable, isAddColumn, column)
+      alterTable(qualifiedTable, isAddColumn, column, defaultValue)
```
[the remark this answers is not in the document]
Comment (dshirish) [reply core/src/main/scala/org/apache/spark/sql/SnappySession.scala]: Corrected the code

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/StoreCallbacksImpl.scala (hunk 1 above)
Comment (dshirish) [reply core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/StoreCallbacksImpl.scala]: Corrected the code

File: core/src/main/scala/org/apache/spark/sql/SnappySession.scala (hunk 2 above)
Comment (dshirish) [reply core/src/main/scala/org/apache/spark/sql/SnappySession.scala]: Have added the implementation

File: core/src/main/scala/org/apache/spark/sql/SnappyDDLParser.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (dshirish) [reply core/src/main/scala/org/apache/spark/sql/SnappyDDLParser.scala]: Have changed the code as per the comment

File: core/src/main/scala/org/apache/spark/sql/SnappyDDLParser.scala [hunk 7]
```diff
@@ -849,8 +859,8 @@ case class DropPolicy(ifExists: Boolean,
 
 case class TruncateManagedTable(ifExists: Boolean, tableIdent: TableIdentifier) extends Command
 
-case class AlterTableAddColumn(tableIdent: TableIdentifier, addColumn: StructField)
-    extends Command
+case class AlterTableAddColumn(tableIdent: TableIdentifier,
+                               addColumn: StructField, defaultValue: String) extends Command
```
[the remark this answers is not in the document]
Comment (dshirish) [reply core/src/main/scala/org/apache/spark/sql/SnappyDDLParser.scala]: Have changed the code as per the comment

**Comment by dshirish** [comment]
@sumwale @kneeraj Have taken care of review comments in this PR from Yogesh.  Please review the changes

**Pull request merged**